Update cheatsheets

sjspielman opened the pull request:
Closes #682 
Related to #667, and possibly even closes it

This PR updates cheatsheets to include some content they were missing and sync up with material swaps. The main content changes include:
- Moved `SingleR` into scAdvanced
- Added `DropletUtils` functions into scIntro
- Added `fs` into intro R

Besides those changes, I didn't find anything else that was obviously missing. I also didn't make any pathway changes related to material swap, since those functions are already in their own cheatsheet.

In addition, I've taken a stab towards #667 with a slightly improved framework cheatsheet creation with an `npm` library. I considered going full GH action here (we could wrap `doctoc` and `mdpdf` in there), but decided to wait for a future PR if we want to take that route! For a nicer review experience in `CONTRIBUTING.md`, recommended to turn off whitespace!

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -113,6 +120,17 @@ Documentation for each of these packages can be accessed by clicking the package
 <br>
 <br>
 
```
There was a bad page break here, so let's make one explicit maybe?  There may be other places we want to do this, but I didn't look fully through all the outputs yet, given some of the other suggested changes.
```suggestion
<div style="page-break-after: always;"></div>
```

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -14,11 +14,13 @@ If you have a different version of R or other R packages, the documentation may
+- [`SingleCellExperiment`, `DropletUtils`, `scran`, and `scater`](#singlecellexperiment-dropletutils-scran-and-scater)
 - [`purrr`](#purrr)
 - [`stringr`](#stringr)
 - [`alevinQC`, `colorblindr`, `Rtsne`, `tibble`](#alevinqc-colorblindr-rtsne-tibble)
```
Noting that we don't use `colorblindr` anymore and probably not `Rtsne` either. This table also has some odd contents, as it includes `S4Vectors::DataFrame`, which should probably go with SCE? And I would put alevinQC with Salmon maybe?

jashapiro reviewed (commented):
Looks good overall, but I added a bit more about installation of Node.js, as well as some places where I think the cheatsheets could use some updates.

sjspielman commented on CONTRIBUTING.md:
```diff
@@ -244,12 +244,24 @@ This workflow then files a PR to `master` with the rendered and live files.
-+ Open the markdown cheatsheet in MacDown, and go to `File` -> `Export` -> `PDF`
-+ Save appropriately and voila!
+Cheatsheets are written in plain markdown and are converted to a shareable PDF format using the `npm` library [`mdpdf`](https://github.com/BlueHatbRit/mdpdf), with the default PDF style.
+You can install this package globally for all users with (or remove the `-g` for local installation) with:
```
> installing into your current directory
🤔 odd indeed!

sjspielman commented on CONTRIBUTING.md:
````diff
@@ -244,12 +244,24 @@ This workflow then files a PR to `master` with the rendered and live files.
+npm install -g mdpdf
+```
+
+In addition, cheatsheet table of contents are created with the `npm` library [`doctoc`](https://github.com/thlorenz/doctoc), which can similarly be globally installed with:
````
I personally am ok with this, _but_ it was my understanding that we were trying to avoid extensions for generating cheatsheets (see my previous attempt to use a VS Code extension for PDF rendering..) and use more standalone/command line tools.

jashapiro replied:
I see a difference between "this may be a convenient alternative with the same effect" and "this is the way it has to be done." If we do make this workflow an action, we can still code all the steps to be sure they are run, with no requirement of VScode.

sjspielman replied:
I definitely see that, but now I'm a little confused on how I should proceed! Are you thinking something like one of these routes? 
- Keep current CONTRIBUTING content, but _also_ suggest that "Markdown All in One" could be used in place of `doctoc` if one were so moved. 
- Keep `mdpdf` but remove `doctoc` and _instead_ recommend that the use of "Markdown All in One" (which I just looked up, does HTML but not PDF conversion) for TOCs
- Focus on the "convenient alternatives" aka VS code extensions for both TOC and PDF rendering. Then later if we want to include an action for rendering, we can write a script to use `doctoc` and `mdpdf` or similar.

sjspielman replied:
Done in https://github.com/AlexsLemonade/training-modules/pull/700/commits/955201b7c993b1c11375324d3bfd39cffee6fef1.

sjspielman commented:
Thanks for helping to catch all the missing functions, @jashapiro ! Just noting I'm going to circle back to this when https://rdrr.io/ isn't down (as it appears to be now, ugh) and I can confidently get the relevant docs links.

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -37,23 +36,35 @@ Read the Base [`R` documentation](https://rdrr.io/r/).
+
+<div style="page-break-after: always;"></div>
+
+### `SingleCellExperiment`, `S4Vectors`, `txmimeta`, and `DropletUtils`
```
Do we need S4vectors in the header? I am not sure we do, though I can see the argument for keeping it.
```suggestion
### `SingleCellExperiment`, `txmimeta`, and `DropletUtils`
```

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
````diff
@@ -89,35 +116,46 @@ Read the [`scater` package documentation](https://rdrr.io/bioc/scater/), and a [
+```
 
-<div style="page-break-after: always;"></div>
+<br>
````
This is ending up splitting between pages anyway, so back to a page break?
```suggestion
<div style="page-break-after: always;"></div>
```

jashapiro commented on module-cheatsheets/intro-to-R-tidyverse-cheatsheet.md:
```diff
@@ -109,14 +109,17 @@ Additional vignettes are available from the "Articles" dropdown menu on this web
 
 
-#### `readr`, `tibble` and `tidyr`
+#### `readr`, `fs`, `tibble` `tidyr`
```
Another page break
```suggestion
<div style="page-break-after: always;"></div>

#### `readr`, `fs`, `tibble` `tidyr`
```

jashapiro commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
Line 45: delete `<br>`

jashapiro commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
```diff
@@ -65,18 +64,6 @@ Read the [`miQC` package documentation](https://rdrr.io/github/greenelab/miQC/),
 | `miQC`| [`plotModel()`](https://rdrr.io/github/greenelab/miQC/man/plotModel.html)| Plot model | `miQC::plotMetics()` with the `miQC` fitted model overlaid |
 | `miQC`| [`plotFiltering()`](https://rdrr.io/github/greenelab/miQC/man/plotFiltering.html)| Plot filtering | Plot percent of mitochondrial reads against the number of unique genes found, coloring points based on whether they will be filtered out or not |
 
-
-## `SingleR`
-
-Read the [`SingleR` package documentation](https://rdrr.io/bioc/SingleR/), and an [e-book on its usage](http://bioconductor.org/books/release/SingleRBook/).
-
-| Library/Package      | Piece of Code      | What it's called    | What it does  |
-|----------------------|--------------------|---------------------|---------------|
-| `SingleR` | [`trainSingleR()`](https://rdrr.io/bioc/SingleR/man/trainSingleR.html) | Train the SingleR classifier | Build a `SingleR` classifier model object from an annotated reference dataset |
-| `SingleR` | [`classifySingleR()`](https://rdrr.io/bioc/SingleR/man/classifySingleR.html) | Classify cells with SingleR | Use a `SingleR` model object to assign cell types to the cells in an `SCE` object |
-| `SingleR` | [`SingleR()`](https://rdrr.io/bioc/SingleR/man/SingleR.html) | Annotate scRNA-seq data | Combines `trainSingleR()` and `classifySingleR()` to assign cell types to an `SCE` object from an annotated reference dataset |
-
-
```
I think it maybe makes sense to have SCE conversion in its own cheatsheet, yes. We already have some of that in the ScPCA docs that we have pointed people to, but I think it does make sense as an "extra". For now we can just have SCE <-> Seurat but there is room to add more later.

jashapiro commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
```diff
@@ -65,18 +64,6 @@ Read the [`miQC` package documentation](https://rdrr.io/github/greenelab/miQC/),
-| `SingleR` | [`SingleR()`](https://rdrr.io/bioc/SingleR/man/SingleR.html) | Annotate scRNA-seq data | Combines `trainSingleR()` and `classifySingleR()` to assign cell types to an `SCE` object from an annotated reference dataset |
-
-
 ## `batchelor` and `harmony`
```
I think a page break here too.
```suggestion
<div style="page-break-after: always;"></div>
## `batchelor` and `harmony`
```

jashapiro reviewed (commented):
Thanks for these updates! I think it is almost there, but I started going through and adding page breaks before realizing that some of those are dependent on prior breaks, so it probably makes sense for you to go through and do all of those all at once. I would only update the sheets for this workshop (& maybe advanced) at this point though.

I think we will probably want to get rid of all `<br>` that may have crept in through time as well.

sjspielman commented:
Noting that I've separated out SCE/Seurat into its own cheatsheet (any naming/textual recommendations there?), and I also added it to the repo's README along (along with scAdvanced which was missing). The new SCE/Seurat cheatsheet is also linked from the scAdvanced cheatsheet, _but the links don't exist yet._ We can confidently anticipate they will exist when this PR is merged!

I've cleaned up the line breaks and removed `<br>`'s. As part of my journey to whitespace greatnesss, I combined `head()`/`tail()` into one row in the intro cheatsheet, which seemed reasonable anyways.

jashapiro commented on CONTRIBUTING.md:
```diff
@@ -244,12 +244,29 @@ This workflow then files a PR to `master` with the rendered and live files.
+* Navigate in terminal to the `module-cheatsheets` directory
+* Run `doctoc` on the file to update its table of contents: `doctoc cheatsheet-file.md`
+  * Alternatively, you can use a tool such as the VS Code Extension [Markdown All in One](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one).
+  If you choose this route, ensure that the table of contents is not duplicated.
```
I'm pretty sure the way this happened is that Markdown All in One does not use HTML comment delimiters and instead identifies the TOC through simple heuristics. So I probably created a TOC using that, so there were no HTML comments, then somebody ran `doctoc` which didn't find its comments, so created a new TOC.

sjspielman replied:
>  then somebody ran `doctoc` which didn't find its comments, so created a new TOC.

Ah yes! It was me 😬

jashapiro commented on README.md:
```diff
@@ -37,5 +37,7 @@ These cheatsheets can also be found in the folder named [module-cheatsheets](htt
 - [RNA seq cheatsheet](https://github.com/AlexsLemonade/training-modules/blob/master/module-cheatsheets/RNA-seq-cheatsheet.md)
 - [scRNA-seq cheatsheet](https://github.com/AlexsLemonade/training-modules/blob/master/module-cheatsheets/scRNA-seq-cheatsheet.md)
+- [Advanced scRNA-seq cheatsheet](https://github.com/AlexsLemonade/training-modules/blob/master/module-cheatsheets/scRNA-seq-advanced-cheatsheet.md)
+- [`SCE` and `Seurat` objects cheatsheet](https://github.com/AlexsLemonade/training-modules/blob/master/module-cheatsheets/SCE-Seurat-cheatsheet.md)
```
I think maybe something a bit more generic? This would give us space to add in the `scverse` content later.
```suggestion
- [Converting to/from `SCE` objects cheatsheet](https://github.com/AlexsLemonade/training-modules/blob/master/module-cheatsheets/SCE-Seurat-cheatsheet.md)
```

jashapiro commented on module-cheatsheets/SCE-Seurat-cheatsheet.md:
```diff
@@ -0,0 +1,104 @@
+| Feature (gene)-level metadata | `rowData(sce_object)` | `seurat_obj[["RNA"]]@meta.features`|
+| Miscellaneous additional metadata | `metadata(sce_object)` | `seurat_obj@misc`|
+
+Below, we provide some code examples below for how you can accomplish these conversions.
```
```suggestion
We provide some code examples below for these conversions below.
```

jashapiro commented on module-cheatsheets/SCE-Seurat-cheatsheet.md:
````diff
@@ -0,0 +1,104 @@
+seurat_object <- Seurat::as.Seurat(sce_object, data = NULL)
+```
+
+<div style="page-break-after: always;"></div>
+
+### Approaches from `ScPCA`
+
+In addition, this [documentation from the `ScPCA`](https://scpca.readthedocs.io/en/latest/faq.html#what-if-i-want-to-use-seurat-instead-of-bioconductor) introduces how to convert `SCE` objects to `Seurat` objects.
+Although this documentation was written for `ScPCA` datasets, the steps generally apply to any `SCE` object.
+It's worth noting that the example code provided at that link will only retain a single assay (raw `"counts"`) in the new `SCE` object, and it will not retain reduced dimension representations (e.g., PCA or UMAP).
+Therefore, this example code is mostly useful at the early stages of processing before you have performed normalization and calculated reduced dimensions.
````
I know I suggested adding this, but I think it may be a bit out of date. I think the `Seurat::as.Seurat()` function should probably be preferred most of the time. My main thought for including this was for conversion to scanpy/scverse, but I think we need to revisit those as well. So maybe we don't include this until we update those.

sjspielman replied:
You mean this whole "approaches from ScPCA" section, right? Just confirming!

jashapiro replied:
Yes, the whole section can go for now!

jashapiro commented on module-cheatsheets/intro-to-R-tidyverse-cheatsheet.md:
```diff
@@ -109,14 +109,19 @@ Additional vignettes are available from the "Articles" dropdown menu on this web
+Read the [`fs` package documentation](https://fs.r-lib.org/).
+Read the [`tibble` package documentation](https://tibble.tidyverse.org/) and a [vignette on its usage](https://tibble.tidyverse.org/articles/tibble.html).
 Read the [`tidyr` package documentation](https://tidyr.tidyverse.org/) and a [vignette on its usage](https://tidyr.tidyverse.org/articles/tidy-data.html).
+https://fs.r-lib.org/reference/index.html
```
I'm assuming this snuck in accidentally. 
```suggestion
```

(I don't think there is any need to include the `fs` comparison article: https://fs.r-lib.org/articles/function-comparisons.html

sjspielman replied:
yep, accident..

jashapiro reviewed (approved):
Looks good, but a couple little tweaks/fixes and rerender still linger. I don't think I need to see it again though.
